src/components/Tooltip/TooltipPortal.tsx

```diff
@@ -0,0 +1,91 @@
+import React, { FC, ReactNode, useEffect, useRef, useState } from 'react'
+import { createPortal } from 'react-dom'
+import styled, { css } from 'styled-components'
+
+import { getTooltipRect } from './tooltipHelper'
+
+type Props = {
+  id: string
+  parentRect: DOMRect
+  children: ReactNode
+  isIcon?: boolean
+  isMultiLine?: boolean
+  horizontal: 'left' | 'center' | 'right'
+  vertical: 'top' | 'middle' | 'bottom'
+}
+
+export const TooltipPortal: FC<Props> = ({
+  id,
+  parentRect,
+  children,
+  isIcon = false,
+  isMultiLine = false,
+  horizontal,
+  vertical,
+}) => {
+  const portalRef = useRef<HTMLDivElement>(null)
+  const [rect, setRect] = useState({
+    top: 0,
+    left: 0,
+    width: 0,
+    height: 0,
+  })
+  useEffect(() => {
+    if (!parentRect || !portalRef.current) {
+      return
+    }
+    const { offsetWidth, offsetHeight } = portalRef.current
+    setRect(
+      getTooltipRect({
+        parentRect,
+        tooltipSize: {
+          width: offsetWidth,
+          height: offsetHeight,
+        },
+        vertical,
+        horizontal,
+        isMultiLine,
+        isIcon,
+        outerMargin: 10,
+      }),
+    )
+  }, [portalRef.current])
```

Comment:
`ref.current` はmutableなのでHookのdependenciesとしては使えない (更新されても再描画されない) らしく、以下の警告が出ます。

> 52:6  warning  React Hook useEffect has missing dependencies: 'horizontal', 'isIcon', 'isMultiLine', 'parentRect', and 'vertical'. Either include them or remove the dependency array. Mutable values like 'portalRef.current' aren't valid dependencies because mutating them doesn't re-render the component  react-hooks/exhaustive-deps

実態として変わる可能性があるの `parentRect` だけっぽいですが、内部的に用いているものは全て入れましょうというのが推奨されているので `[parentRect, vertical, horizontal, isMultiLine, isIcon]` を入れたら安心かなと思います！

Reply:
ありがとうございます。
足りていないdepsを追加しました。

`ref.current`は、初回のuseEffectでtooltipの正確なrectangleが取れないために必要だったのですが、その原因が`createPortal()`を併用していることだったので、`createPortal()`を親側に移動することで初回useEffectでrectangleを取得できるようにして`ref.current`は削除しました。